README.md
# 💀 KDMApp-Mods README 💀

## Info
Here are some additions/mods made to the KDM Helper App found here:
https://kdm-story.site/

----

## Install
_assumes you have the [kdm-sotry app](https://kdm-story.site/) installed_  
Download and extract (overwrite) all files from this repo ([link here](https://github.com/jparkerweb/kdmapp-mods/archive/master.zip)) into the `Kingdom Death 1.5 Story\resources\app` folder.

----
### Discussion/Community
Discuss the `kdm-story` app with the BGG community here:
https://boardgamegeek.com/article/31989187

----

## Mod List

* __Hunt Event Input Pad__  
mouse friendly keypad for entering hunt event numbers  
![hunt event input pad](docs/keypad.gif)

* __Reference Keyboard__
mouse friendly keyboard for entering reference entires
![reference keyboard](docs/reference-keyboard.gif)


----
#### Notes

_The intention is to keep these mods updated and compatible with any changes made to the official app (always assuming the latest version is being used).  Eventually we may be able to merge these changes in once the official app can be contributed to in open source fashion._

_feedback welcome_


CHANGELOG.md
# 💀 KDMApp-Mods Change Log 💀

## v1.0.1
* update keyboard to `qwerty` layout
* adjust numpad position for smaller screens

## v1.0.0
* added mouse friendly number `Keypad` for entering Hunt Event Rolls
* added mouse friendly `Keyboard` for interactive with the `Reference` window


js/about.js
function createAbout() {
	$('#container').append($('<div>',{id:'reference-window-background'}));
	$('#reference-window-background').hide();

	// $('#container').append($('<img id="about_text">About</div>'));
	$('#container').append($('<img id="about_text" src="images/icons/about.png"/>'));
	$('#about_text').hide();
	$('#about_text').delay(1000).fadeIn(2000);

	$('#container').append($('<div>',{id:'reference-window-back0', style:'left:36%;'}));
	$('#reference-window-back0').append($('<img>',{id:'reference-window-back-img', src:'images/reference/reference_back.png'}));
	$('#reference-window-back0').append($('<div>',{id:'reference-window-back'}));

	$('<div><h id="about_header">KDM Story app v.2.2</h><p id="about_p">Created by @prikhodkop</p><p id="about_p">Almost all the art and text presented in the app are the property of Adam Poots. The guy did a great job creating the game, so please pay respect to him and <b>use the app ONLY if you own a copy of the game</b> (no exceptions)!</p><p id="about_p">The app also uses some of content created by Fen, please support him at his: </br> https://www.patreon.com/FenPaints </br>He deservers it!.</p><p id="about_p">Hope you enjoy the app and please send your feedback to me!</p></div>').appendTo('#reference-window-back')

	$('#reference-window-back0').hide();

	$('#about_text').hover(function()
	{
		if(!$(this).hasClass('active'))
		{
			if(!$(this).hasClass('hoverd'))
			{
				$('#reference-window-back0').fadeIn(500);
				$('#reference-window-background').fadeIn(500);
				$(this).toggleClass('active');
				$('#about_text').attr('src', 'images/icons/about_active.png');
			}
		}
	},function()
	{
		$(this).delay(500).removeClass('hoverd')
	});

	$('#about_text').click(function(){
		if(!$(this).hasClass('active'))
		{
			$('#reference-window-back0').fadeIn(500);
			$('#reference-window-background').fadeIn(500);
			$('#about_text').attr('src', 'images/icons/about_active.png');
		} else
		{
			$('#reference-window-back0').fadeOut(500);
			$('#reference-window-background').fadeOut(500);
			$('#about_text').attr('src', 'images/icons/about.png');
		};
		$(this).toggleClass('active');
	});

	$('#reference-window-background').on('click', function(){
		$('#reference-window-back0').fadeOut(500);
		$('#reference-window-background').fadeOut(500);
		$('#about_text').removeClass('active');
		$('#about_text').attr('src', 'images/icons/about.png');
	});

	$('#reference-window-back').on('click', function(){
		$('#reference-window-back0').fadeOut(500);
		$('#reference-window-background').fadeOut(500);
		$('#about_text').removeClass('active');
		$('#about_text').attr('src', 'images/icons/about.png');
	});

	$('#about_header').on('click', function(){
		$('#reference-window-back0').fadeOut(500);
		$('#reference-window-background').fadeOut(500);
		$('#about_text').removeClass('active');
		$('#about_text').attr('src', 'images/icons/about.png');
	});

	$('#about_p').on('click', function(){
		$('#reference-window-back0').fadeOut(500);
		$('#reference-window-background').fadeOut(500);
		$('#about_text').removeClass('active');
		$('#about_text').attr('src', 'images/icons/about.png');
	});

}


js/events.js
var events = {
	'index' :      {'table': 1, 'type': 'index', 'label': 'Intro', 'ltop':'50%', 'lleft':'50%'},
	'hunt' :       {'table': 1, 'type': 'hunt', 'label': 'The Hunt', 'ltop':'70%', 'lleft':'50%', 'music':'music/white secret.mp3', 'speech':"audio/speech/en/hunt.mp3"},
	'settlement' : {'table': 1, 'type': 'settlement', 'label': 'Settlement', 'music':'music/white secret.mp3', 'music_delay':'2000', 'speech':"audio/speech/en/settlement.mp3"},
	'age' :                {'ltop':'27%', 'lleft':'49%'},
	'armored strangers' :  {'music':'music/armored strangers.mp3', 'music_delay':'4000'},
	'birth of a savior' :  {'music':'music/birth of a savior.mp3', 'music_delay':'speech-1000', 'ltop':'15%', 'lleft':'50%'},
	'black harvest' :      {'music':'music/rumbling in the dark.mp3'},
	'blackout' :           {'music':'music/overwhelming darkness.mp3', 'music_delay':'3000', 'ltop':'37%', 'lleft':'50%'},
	'bold' :               {'ltop':'50%', 'lleft':'75%'},
	'bone witch' :         {'music':'music/see the truth.mp3', 'ltop':'50%', 'lleft':'65%'},
	'cooking' :            {'music':'music/cooking.mp3', 'music_delay':'500', 'ltop':'73%', 'lleft':'50%'},
	'crush and devour' :   {'music':'music/crush and devour.mp3'},
	'endless screams' :    {'music':'music/crush and devour.mp3', 'ltop':'19%', 'lleft':'48%'},
	'fetid grotto' :       {'music':'music/overwhelming darkness.mp3', 'ltop':'20%', 'lleft':'30%'},
	'final march' :        {'music':'music/final march.mp3', 'music_delay':'500', 'ltop':'20%', 'lleft':'50%'},
	'game over' :          {'music':'music/game over.mp3', 'ltop':'32%', 'lleft':'47%'},
	'hammer and nail' :    {'music':'music/hammer and nail.mp3'},
	'hands of heat' :      {'music':'music/hands of heat.mp3', 'ltop':'75%', 'lleft':'37%'},
	'herb gathering' :     {'music':'music/herb gathering.mp3', 'hide_label':true},
	'hooded knight' :      {'music':'music/hooded knight.mp3', 'music_delay':'speech-1000'},
	'insight' :            {},
	'intimacy' :           {'music':'music/intimacy.mp3', 'music_delay':'speech-1000', 'ltop':'22%', 'lleft':'75%'},
	'kings curse' :        {'music':'music/kings curse.mp3', 'ltop':'88%', 'lleft':'50%'},
	'kings step' :         {'music':'music/kings step.mp3', 'music_delay':'speech+500', 'ltop':'67%', 'lleft':'53%'},
	'lantern research' :   {},
	'legendary lungs' :    {'music':'music/legendary lungs.mp3', 'music_delay':'500'},
	'legendary monsters' : {'music':'music/legendary monsters.mp3', 'hide_label':true},
	'melting horror' :     {'music':'music/crush and devour.mp3'},
	'mineral gathering' :  {'music':'music/mineral gathering.mp3', 'hide_label':true},
	'overwhelming darkness' : {'music':'music/overwhelming darkness.mp3', 'music_delay':'3000'},
	'oxidation' :            {'ltop':'75%', 'lleft':'35%'},
	'phoenix feather' :      {'music':'music/legendary monsters.mp3'},
	'principle conviction' : {'music':'music/principle.mp3', 'music_delay':'speech-1000'},
	'principle death' :      {'music':'music/principle.mp3', 'music_delay':'speech-1000'},
	'principle new life' :   {'music':'music/principle.mp3', 'music_delay':'speech-1000', 'ltop':'85%', 'lleft':'50%'},
	'principle society' :    {'music':'music/principle.mp3', 'music_delay':'speech-1000', 'ltop':'20%', 'lleft':'65%'},
	'regal visit' :          {'music':'music/regal visit.mp3', 'music_delay':'speech-1000', 'ltop':'65%', 'lleft':'50%'},
	'returning survivors' :  {},
	'rumbling in the dark' : {'type':'video', 'music':'music/rumbling in the dark.mp3'},
	'run away' :             {'music':'music/run away.mp3'},
	'secret meeting' :       {'music':'music/secret meeting.mp3', 'music_delay':'500', 'ltop':'35%', 'lleft':'40%'},
	'see the truth' :        {'music':'music/see the truth.mp3'},
	'showdown butcher' :     {'music':'music/showdown butcher.mp3'},
	'showdown dung beetle knight' : {'music':'music/showdown butcher.mp3', 'ltop':'5%', 'lleft':'55%'},
	'showdown gold smoke knight' :  {'music':'music/showdown gold smoke knight.mp3', 'ltop':'87%', 'lleft':'80%'},
	'showdown gorm' :        {'music':'music/showdown gorm.mp3'},
	'showdown kings man' :   {'music':'music/showdown kings man.mp3', 'ltop':'68%', 'lleft':'63%'},
	'showdown phoenix' :     {'music':'music/showdown phoenix.mp3'},
	'showdown screaming antelope' : {'music':'music/showdown query.mp3', 'ltop':'15%', 'lleft':'57%'},
	'showdown the hand' :    {'music':'music/showdown the hand.mp3'},
	'showdown watcher' :     {'music':'music/showdown watcher.mp3', 'ltop':'15%', 'lleft':'50%'},
	'showdown white lion' :  {'music':'music/showdown query.mp3', 'ltop':'65%', 'lleft':'75%'},
	'spelunking of death' :  {'music':'music/spelunking of death.mp3', 'ltop':'18%', 'lleft':'63%'},
	'the approaching storm' : {'music':'music/the approaching storm.mp3', 'music_delay':'500', 'ltop':'20%', 'lleft':'72%'},
	'underground sow' :      {'music':'music/rumbling in the dark.mp3', 'ltop':'20%', 'lleft':'40%'},
	'watched' :              {'music':'music/watched.mp3', 'ltop':'15%', 'lleft':'50%'},
	'white secret' :         {'music':'music/white secret.mp3'},
	'white speaker' :        {'music':'music/white speaker.mp3', 'music_delay':'speech-1000', 'ltop':'8%', 'lleft':'66%'},
	'zero presence' :        {'music':'music/overwhelming darkness.mp3'}
}

var lang = 'en'

function Event(
	id,
	type="image",
	campaign="",
	expansion="",
	music="theme.mp3", // files are assumed to be in audio folder
	music_delay="speech-2000",
	speech="",
	label="###",
	ltop="75%",
	lleft="50%",
	table=2, //1 -displayed with larger font on top row or 2 - smaller font big table
	hide_label = false
	) {
		this.type = type;
		this.campaign = campaign;
		this.expansion = expansion;
		this.music = "audio/"+music;
		this.music_delay = music_delay;

		if (speech == "") {
			this.speech = "audio/speech/"+lang+"/"+id+".mp3"
		} else {
			this.speech = "audio/"+speech
		}

		if (label == "###") {
			let formatted_label = id.replace('showdown', 'showdown:').replace('principle', 'principle:')
			formatted_label = titleCase(formatted_label)
			this.label = formatted_label;
		} else {
			this.label = label;
		}

		this.ltop = ltop;
		this.lleft = lleft;
		this.table = table;
		this.hide_label = hide_label;
}

function create_events_table(events) {

	let events_table = {};

	let event_ids = Object.keys(events)

	for (let i = 0; i < event_ids.length; i++) {
		events_table[event_ids[i]] = new Event(event_ids[i]);

		let event = events[event_ids[i]]
		for (let property in event) {
			if (event.hasOwnProperty(property)) {
				if (property == 'music') {
					events_table[event_ids[i]][property] = './audio/' + event[property];
				} else {
					events_table[event_ids[i]][property] = event[property];
				}

					// console.log('Set property for '+event_ids[i]+':');
					// console.log(property+' : '+ event[property]);
			}
		}
	}

	return events_table
};


// GLOBAL TABLE OF ALL EVENTS TO BE DISPLAYED AT MENU
// ##################
var events_table = create_events_table(events);
// ##################


function titleCase(str) {
	let splitStr = str.toLowerCase().split(' ');
	for (let i = 0; i < splitStr.length; i++) {
			splitStr[i] = splitStr[i].charAt(0).toUpperCase() + splitStr[i].substring(1);
	}
	return splitStr.join(' ');
}

function createToc(col_len=5) {

	// PREPARE LIST OF OBJECTS FOR TABLE 1 AND TABLE 2
	// ####################
	let events_table_ids = Object.keys(events_table)

	let table_1_ids = [];
	let table_2_ids = [];

	for (let i = 0; i < events_table_ids.length; i++) {
		if (events_table[events_table_ids[i]].table == 1) {
			table_1_ids.push(events_table_ids[i]);
		} else if (events_table[events_table_ids[i]].table == 2) {
			table_2_ids.push(events_table_ids[i]);
		} else {
			throw 'Wrong table type ('+events_table[events_table_ids[i]].table+') for event '+events_table_ids[i]+'!!!'
		}

	}

	table_2_ids.sort();
	// #####################

	let container = document.getElementById("menu")

	let width1 = Math.floor(100/table_1_ids.length)
	let width2 = Math.floor(100/col_len)

	// CREATE TABLE 1
	let tbl1  = document.createElement('table');
	// tbl1.style.width  = '80%';
	// tbl1.style.top  = '3%';
	tbl1.setAttribute("id", "menu_table1");
	tbl1.className = "top";
	tbl1.style.cssText += 'width:80%;top:3%;left:10%;'

	let tbdy1 = document.createElement('tbody');

	// JUST ONE ROW FOR TABLE 1
	// let tr = tbl1.insertRow();
	let tr = document.createElement('tr');
	// tr.className = "spaceUnder";
	tr.style.cssText += 'width: 100%; position:static;'

	for(let j = 0; j < table_1_ids.length; j++){
		// let td = tr.insertCell();
		let td = document.createElement('td');
		td.style.cssText += "width:"+width1+"%;";
		// let a1 = td;

		let a1 = document.createElement('div');
		a1.setAttribute("id", "menu_item");
		a1.className = "hunt_menu";
		// a.setAttribute("href", events_table[rows[i][j]].type+'.html');
		// a.setAttribute("onclick", "setTransition('"+rows[i][j]+"')");
		a1.setAttribute("target", table_1_ids[j])
		a1.style.cssText += "font-size: 2.5em;width:100%;position:static; margin:auto;"

		a1.innerHTML += events_table[table_1_ids[j]].label;

		td.appendChild(a1)

		tr.appendChild(td)
	}

	tbdy1.appendChild(tr)
	tbl1.appendChild(tbdy1)
	container.appendChild(tbl1);

	// CREATE TABLE 2

	let tbl2  = document.createElement('table');
	// tbl2.style.width  = '90%';
	// tbl2.style.top  = '12%';
	tbl2.setAttribute("id", "menu_table2");
	tbl2.style.cssText += 'width:90%;top:12%;left:5%;'

	let row_len = Math.ceil(table_2_ids.length / col_len)

	let rows = []

	for (let i = 0; i < row_len; i++) {
		rows.push([])
	};

	let ctr = 0
	let ro_idx = 0
	for (let i = 0; i < table_2_ids.length; i++) {
		if (ctr >= row_len) {
			ctr = 0
			// cl_idx = cl_idx + 1
			ro_idx = 0
		}
		rows[ro_idx].push(table_2_ids[i])
		ro_idx = ro_idx + 1
		ctr = ctr + 1
	};

	// console.log(rows);

	for(let i = 0; i < rows.length; i++){
		let tr = tbl2.insertRow();
		// tr.className = "spaceUnder";
		for(let j = 0; j < rows[i].length; j++){
			let td = tr.insertCell();
			td.style.cssText = "width:"+width2+"%;height:5em;position:static;";

			let a2 = document.createElement('div');
			a2.setAttribute("id", "menu_item");
			// a.setAttribute("href", events_table[rows[i][j]].type+'.html');
			// a.setAttribute("onclick", "setTransition('"+rows[i][j]+"')");
			a2.setAttribute("target", rows[i][j])
			a2.style.cssText += "width:100%;position:static; margin:auto;"

			a2.innerHTML += events_table[rows[i][j]].label;

			td.appendChild(a2)


		}
	}

	container.appendChild(tbl2);

}


js/hunt_events.js

var path_to_quary_events = 'images/hunt/quary_events/'

var quary_events = [
	'dung beetle knight 1.jpg',
	'dung beetle knight 2.jpg',
	'dung beetle knight 3.jpg',
	'dung beetle knight 4.jpg',
	'dung beetle knight 5.jpg',
	'dung beetle knight 6.jpg',
	'dung beetle knight 7.jpg',
	'gorm 1.jpg',
	'gorm 2.jpg',
	'gorm 3.jpg',
	'gorm 4.jpg',
	'gorm 5.jpg',
	'gorm 6.jpg',
	'gorm 7.jpg',
	'lion god 1.jpg',
	'lion god 2.jpg',
	'lion god 3.jpg',
	'lion god 4.jpg',
	'lion god 5.jpg',
	'lion god 6.jpg',
	'lion god 7.jpg',
	'lion god 8.jpg',
	'phoenix 1.jpg',
	'phoenix 2.jpg',
	'phoenix 3.jpg',
	'phoenix 4.jpg',
	'phoenix 5.jpg',
	'phoenix 6.jpg',
	'phoenix 7.jpg',
	'phoenix 8.jpg',
	'phoenix 9.jpg',
	'screaming antelope 1.jpg',
	'screaming antelope 2.jpg',
	'screaming antelope 3.jpg',
	'screaming antelope 4.jpg',
	'screaming antelope 5.jpg',
	'screaming antelope 6.jpg',
	'screaming antelope 7.jpg',
	'screaming antelope 8.jpg',
	'screaming antelope 9.jpg',
	'sunstalker 1.jpg',
	'sunstalker 2.jpg',
	'sunstalker 3.jpg',
	'sunstalker 4.jpg',
	'sunstalker 5.jpg',
	'sunstalker 6.jpg',
	'sunstalker 7.jpg',
	'sunstalker 8.jpg',
	'white lion 1.jpg',
	'white lion 2.jpg',
	'white lion 3.jpg',
	'white lion 4.jpg',
	'white lion 5.jpg',
	'white lion 6.jpg',
	'white lion 7.jpg'
]


function get_sequence(name) {

		let sequence = []

		for (let i = 0; i < quary_events.length; i ++) {
			if (quary_events[i].includes(name.replace('showdown ',''))) {
				sequence.push(path_to_quary_events+quary_events[i])
			}
		}

		return shuffle(sequence);
}

function next_event(sequence, idx) {
	if (idx >= sequence.length) {
		idx = 0
		sequence = shuffle(sequence)
	}

	let next = sequence[idx]
	idx += 1;

	return next, sequence, idx
};

function idx_event(sequence, idx) {

	return sequence[idx]
};



function shuffle(a) {
		let j, x, i;
		for (i = a.length - 1; i > 0; i--) {
				j = Math.floor(Math.random() * (i + 1));
				x = a[i];
				a[i] = a[j];
				a[j] = x;
		}
		return a;
}


js/menu.js
function createMenuButton() {
	// <a href="javascript:void(0)" id="menu-toggle-wrapper">
	//   <div id="menu-toggle"></div>
	// </a>
	//
	// <div id="menu" style="opacity:.9;">
	//   <img id="menu_img" src="images/back.jpg"/>
	// </div>
	$('#container').append($('<a>',{href:'javascript:void(0)',id:'menu-toggle-wrapper'}));
	$('#container').append($('<div>',{style:'opacity:.9;',id:'menu'}));

	$('#menu-toggle-wrapper').append($('<div>',{id:'menu-toggle'}));
	$('#menu').append($('<img>',{id:'menu_img', src:'images/back.jpg'}));

	$('#menu').hide()

	$('#menu-toggle-wrapper').hover(function()
	{
		if(!$(this).hasClass('active'))
		{
			if(!$(this).hasClass('hoverd'))
			{
				$("#menu").fadeIn(500);
				$(this).toggleClass('active');
			}
		}
	},function()
	{
		$(this).delay(500).removeClass('hoverd')
	});

	$('#menu-toggle-wrapper').click(function(){
		if(!$(this).hasClass('active'))
		{
			$("#menu").fadeIn(500);
		} else
		{
			$("#menu").hide();
		};
		$(this).toggleClass('active');
	});
}

function createSevereTables() {

	$('#container').append($('<img>',{id:'severe-table',src:''}));
	addLocationTable('brain', '11%');
	addLocationTable('head', '18%');
	addLocationTable('arms', '25%');
	addLocationTable('body', '32%');
	addLocationTable('waist', '39%');
	addLocationTable('legs', '46%');

	// $('#container').append($('<div>',{id:'severe-background'}));

	// $(document).on(
	// {
	//   mouseleave: function() {
	//       $(this).removeClass('active');
	//       hideLocationTable();
	//   },
	// },
	// 'aaa')​

}

function addLocationTable(location, top) {
	$('#container').append($('<img>',{class:location,id:'severe',src:'images/icons/'+location+'.png', style:'top:'+top+';'}))
	$('#container').append($('<img>',{class:location,id:'severe-table',src:'images/severe injuries/'+location+'.png'}))
	$('#severe.'+location).hide();
	$('#severe-table.'+location).hide();
	$('#severe.'+location).delay(1000).fadeIn(2000);



	$('#severe.'+location).hover(function()
	{
		if(!$(this).hasClass('active'))
		{
			if(!$(this).hasClass('hoverd'))
			{
				// $("#menu").fadeIn(500);
				showLocationTable($(this).attr('class'))
				$(this).toggleClass('active');
			}
		}
	},function()
	{
		$(this).delay(500).removeClass('hoverd')
	});

	$('#severe.'+location).click(function(){
		if(!$(this).hasClass('active'))
		{
			showLocationTable($(this).attr('class'))
		} else
		{
			hideLocationTable()
		};
		$(this).toggleClass('active');
	});

	$(document).on({
		mouseenter: function() {
				$("#divtwo").css("background-color", "yellow");
		},
		mouseleave: function() {
				$(this).removeClass('active');
				hideLocationTable($(this).attr('class'));
		}
	}, "#severe");
}

function showLocationTable(location) {
	// $('#severe-background').fadeIn(500);
	// $('#severe-table').hide();
	// $('#severe-table').attr('src', 'images/severe injuries/'+location+'.png');
	$('#severe-table.'+location).delay(100).fadeIn(200);
	// $('#severe-table').slideLeft(1000);
}

function hideLocationTable(location) {
	$('#severe-table.'+location).fadeOut(100);
	// $('#severe-background').delay(500).fadeOut(00);
	// $('#severe-table').slideRight(1000);
}

function createReference() {
	$('#container').append($('<div>',{id:'reference-window-background'}));
	$('#reference-window-background').hide();

	$('#container').append($('<img>',{id:'reference',src:'images/icons/reference.png'}));
	$('#reference').hide();
	$('#reference').delay(1000).fadeIn(2000);

	$('#container').append($('<div>',{id:'reference-window-back0'}));
	$('#reference-window-back0').append($('<img>',{id:'reference-window-back-img', src:'images/reference/reference_back.png'}));
	// $('#reference-window-back').append('<label for="reference-window">Terms:</label>')
	$('#reference-window-back0').append('<select id="reference-window" class="reference-window" placeholder="Type term..."></select>')
	$('#reference-window-back0').append($('<div>',{id:'reference-window-back'}));
	$('#reference-window-back0').append($('<div>',{id:'reference-window-keypad'}));
	populateRefKeypad();

	$('#reference-window-back0').hide();

	$('#reference-window-back').append($('<img>',{id:'glossary-symbols',src:'images/reference/symbols.png'}));

	let all_option = get_all_options();

	console.log(all_option)

	var selectize = $('#reference-window').selectize({
		options: all_option,
		optgroups: [
				{value: 'random draws', label: 'Random Draws'},
				{value: 'settlement events', label: 'Settlement Events'},
				{value: 'glossary', label: 'Glossary'},
				{value: 'abilities', label: 'Abilities'},
				{value: 'fightning arts', label: 'Fightning Arts'},
				{value: 'disorders', label: 'Disorders'},
				{value: 'secret fightning arts', label: 'Secret Fightning Arts'},

		],
		optgroupField: 'class',
		labelField: 'name',
		searchField: ['name'],
		maxItems: 10,
		plugins: ['remove_button','silent_remove'],
		// render: {
		//     optgroup_header: function(data, escape) {
		//         return '<div class="optgroup-header">' + escape(data.label) + ' <span class="scientific">' + escape(data.label_scientific) + '</span></div>';
		//     }
		// }
		hideSelected: true,
		closeAfterSelect: true,
		sortField: [
				{
						field: 'name',
						direction: 'asc'
				},
				{
						field: '$score'
				}
		],
		// ignoreFocusOpen: true,
		onItemRemove: function(values) {
				// return confirm(values);
				console.log(values);
				$('#reference-data.'+adapt_name(values)).fadeOut(500,function(){
					$(this).css({"visibility":"hidden",display:'block'}).slideUp();
				});
				// $('.selectize-dropdown').css("display", "none");

				setTimeout(function(){$('#reference-data.'+adapt_name(values)).remove()}, 1500);
				// $('#reference-window').selectize.close();
				// $('#reference-window').setValue('Type here...');
				setTimeout(function(){
					if (!$('.selectize-input').hasClass('has-items')) {
						$('#glossary-symbols').fadeIn(500);
					}
				}, 100);
		},

		onItemAdd: function(values, item) {
				// return confirm(item);
				// console.log(item);

				if (is_random_draw(values)) {
					let draws = get_random_draws(values);
					console.log(values)
					console.log('Draws:')
					console.log(draws)


					selectize.removeItem(values, true);
					$('#glossary-symbols').hide();
					for (let i = 0; i < draws.length; i++) {
						console.log(draws[i])
						setTimeout(function(){selectize.addItem(draws[i], false);$('#glossary-symbols').hide();}, 500*i);
					}

				} else {
					// $('#reference-window-back').append('<div id="reference-data" class="'+adapt_name(values)+'">'+get_representation(values)+'<br/><br/></div>')
					// $('#reference-data.'+adapt_name(values)).hide();
					// $('#reference-data.'+adapt_name(values)).delay(50).fadeIn(500);
					show_element(values)
				}
		},

		onDropdownOpen: function($dropdown) {
			if (!this.lastQuery.length) {
				this.close();
			} else {
				$('#glossary-symbols').hide();
			};
		},

		onDropdownClose: function($dropdown) {
			if (!$('.selectize-input').hasClass('has-items')) {
				$('#glossary-symbols').fadeIn(500);
			}
		},
	})[0].selectize;

	// selectize.on('item_remove', function() {
	//   selectize.close();
	// });

	selectize.$control_input.on('keydown', function(e) {
		if (e.keyCode == 8 && selectize.$control_input.val().length < 2) {
			selectize.close();
		}
	});

	function populateRefKeypad() {
		var refPadArr = ['row-start','q','w','e','r','t','y','u','i','o','p','row-end','row-start','a','s','d','f','g','h','j','k','l','row-end','row-start','z','x','c','v','b','n','m','row-end','row-start','space','clear','random','row-end']
		var txtPad = "<input type='hidden' id='refPadHiddenValue' />"

		var i;
		for (i = 0; i < refPadArr.length; i++) { 
			if (refPadArr[i] === 'row-start') { txtPad = txtPad + "<div class='refpad--row'>" }
			else if (refPadArr[i] === 'row-end') { txtPad = txtPad + "</div>" }
			else {
				if (refPadArr[i].length > 0) {
					txtPad = txtPad + "<div class='refpad__pad -" + refPadArr[i] + "' onClick='refPadEntry(\"" + refPadArr[i] + "\")'>" + refPadArr[i] + "</div>"
				} else {
					txtPad = txtPad + "<div class='refpad__pad' onClick='refPadEntry(\"" + refPadArr[i] + "\")'>" + refPadArr[i] + "</div>"
				}
			}
		}

		let refKeypad = document.getElementById("reference-window-keypad")
		refKeypad.innerHTML = txtPad
	}

	$('#reference').hover(function()
	{
		if(!$(this).hasClass('active'))
		{
			if(!$(this).hasClass('hoverd'))
			{
				$('#reference-window-back0').fadeIn(500);
				$('#reference-window-background').fadeIn(500);
				$('#reference').attr('src', 'images/icons/reference_active.png');
				selectize.focus();
				$(this).toggleClass('active');
			}
		}
	},function()
	{
		$(this).delay(500).removeClass('hoverd')
	});

	$('#reference').click(function(){
		if(!$(this).hasClass('active'))
		{
			$('#reference-window-back0').fadeIn(500);
			$('#reference-window-background').fadeIn(500);
			$('#reference').attr('src', 'images/icons/reference_active.png');
			selectize.focus();
		} else
		{
			$('#reference-window-back0').fadeOut(500);
			$('#reference-window-background').fadeOut(500);
			$('#reference').attr('src', 'images/icons/reference.png');
		};
		$(this).toggleClass('active');
	});

	$('#reference-window-background').on('click', function(){
		$('#reference-window-back0').fadeOut(500);
		$('#reference-window-background').fadeOut(500);
		$('#reference').attr('src', 'images/icons/reference.png');
		$('#reference').removeClass('active');
	});

}

function refPadEntry(refValue) {
	var refInput = document.getElementById("reference-window-selectized")
	var refHiddenInput = document.getElementById("refPadHiddenValue")
	// console.log("refInput.value", refInput.value)
	// console.log("refHiddenInput.value", refHiddenInput.value)
	
	if (refValue === "space") {
		refHiddenInput.value = refHiddenInput.value + " "	
	} else if (refValue === "random") {
		refHiddenInput.value = "random"
	} else if (refValue === "clear") {
		refHiddenInput.value = ""
	} else {
		refHiddenInput.value = refHiddenInput.value + refValue	
	}

	refInput.value = refHiddenInput.value
	refInput.focus()
}

function adapt_name(name) {
	return name.replace(/ /g, '_').replace(/\(/g, 'A').replace(/\)/g, 'A').replace(/\//g, 'A').replace(/\"/g, 'A').replace(/,/g, 'A').replace(/&/g, 'n')
}

function show_element(value) {
	$('#reference-window-back').append('<div id="reference-data" class="'+adapt_name(value)+'">'+get_representation(value)+'<br/><br/></div>')
	$('#reference-data.'+adapt_name(value)).hide();
	$('#reference-data.'+adapt_name(value)).delay(50).fadeIn(500);
}
